vpr/src/route/rr_graph.cpp

```diff
@@ -3198,26 +3299,44 @@ static vtr::NdMatrix<int, 5> alloc_and_load_pin_to_seg_type(const e_pin_type pin
         if (Type->is_ignored_pin[pin])
             continue;
 
-        for (int width = 0; width < Type->width; ++width) {
-            for (int height = 0; height < Type->height; ++height) {
-                for (e_side side : SIDES) {
-                    if (Type->pinloc[width][height][side][pin] == 1) {
-                        dir_list[width][height][side][num_dir[width][height][side]] = pin;
-                        num_dir[width][height][side]++;
+        for (auto type_layer_index : type_layer) {
+            for (int width = 0; width < Type->width; ++width) {
+                for (int height = 0; height < Type->height; ++height) {
+                    for (e_side side : SIDES) {
+                        if (Type->pinloc[width][height][side][pin] == 1) {
+                            //connect the pin to its own layer tracks
+                            dir_list[width][height][type_layer_index][side][num_dir[width][height][type_layer_index][side]] = pin;
+                            num_dir[width][height][type_layer_index][side]++;
+                            //connect the pin to the layer that is specified in the arch file by "layer_offset"
+                            int pin_layer = type_layer_index + Type->pin_layer_offset[pin];
+                            if (pin_layer >= grid.get_num_layers()) { //pin layer is invalid, wrap it around
```

Comment:
Based on our previous discusstion, I think this implementation may change. Thus, wrapping it in a helper function would make future changes easier.